strided_view() + flatten() + assignment = compilation failure

I'm trying to construct a strided view, flatten it, and then assign to that view. This results in a compilation failure: 

(same results if you remove the column major pieces)

(same results if you assign something like `xt::xarray<int> value = {1};` to the view, rather than just `1`)

```cpp
// [[Rcpp::depends(xtensor)]]
// [[Rcpp::plugins(cpp14)]]

#include "xtensor/xarray.hpp"
#include "xtensor/xadapt.hpp"
#include "xtensor/xstrided_view.hpp"
#include "xtensor/xio.hpp"

// [[Rcpp::export]]
void strided_view_flatten_assign() {
  xt::xarray<int, xt::layout_type::column_major> x = {{1, 2, 3}, {4, 5, 6}};
  
  auto x_view = xt::strided_view(x, {xt::range(0, 1), xt::range(0, 1)});
  
  auto x_flat = xt::flatten<xt::layout_type::column_major>(x_view);
  
  x_flat = 1;
  
  std::cout << "out" << std::endl << x_flat << std::endl;
}
```

```
test.cpp:17:10: error: no viable overloaded '='
  x_flat = 1;
  ~~~~~~ ^ ~
/Library/Frameworks/R.framework/Versions/3.6/Resources/library/xtensor/include/xtensor/xtensor.hpp:239:26: note: candidate function not viable: no known conversion from 'int' to 'const xt::xtensor_adaptor<xt::xiterator_adaptor<xt::xiterator<xt::xstepper<xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major>, xt::xiterator<xt::xstepper<const xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major> >, 1, xt::layout_type::dynamic, xt::xtensor_expression_tag>' for 1st argument
        xtensor_adaptor& operator=(const xtensor_adaptor&);
                         ^
/Library/Frameworks/R.framework/Versions/3.6/Resources/library/xtensor/include/xtensor/xtensor.hpp:242:26: note: candidate function not viable: no known conversion from 'int' to 'xt::xtensor_adaptor<xt::xiterator_adaptor<xt::xiterator<xt::xstepper<xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major>, xt::xiterator<xt::xstepper<const xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major> >, 1, xt::layout_type::dynamic, xt::xtensor_expression_tag>' for 1st argument
        xtensor_adaptor& operator=(xtensor_adaptor&&);
                         ^
/Library/Frameworks/R.framework/Versions/3.6/Resources/library/xtensor/include/xtensor/xtensor.hpp:243:26: note: candidate function not viable: no known conversion from 'int' to 'xt::xtensor_adaptor<xt::xiterator_adaptor<xt::xiterator<xt::xstepper<xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major>, xt::xiterator<xt::xstepper<const xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major> >, 1, xt::layout_type::dynamic, xt::xtensor_expression_tag>::temporary_type' (aka 'xtensor_container<xt::uvector<int, std::__1::allocator<int> >, 1UL, (xt::layout_type)0, xt::xtensor_expression_tag>') for 1st argument
        xtensor_adaptor& operator=(temporary_type&&);
                         ^
/Library/Frameworks/R.framework/Versions/3.6/Resources/library/xtensor/include/xtensor/xtensor.hpp:246:26: note: candidate template ignored: could not match 'xexpression<type-parameter-0-0>' against 'int'
        xtensor_adaptor& operator=(const xexpression<E>& e);
                         ^
1 error generated.
make: *** [test.o] Error 1
```

Also technically the original example still fails to compile, assigning `x_flat = 1` doesn't work because the `1` isn't wrapped in an `xtensor` style object. Basically you still can't assign a scalar value. (it doesn't affect me, but just an fyi)

<details>

```
test.cpp:17:10: error: no viable overloaded '='
  x_flat = 1;
  ~~~~~~ ^ ~
/Library/Frameworks/R.framework/Versions/3.6/Resources/library/xtensor/include/xtensor/xtensor.hpp:239:26: note: candidate function not viable: no known conversion from 'int' to 'const xt::xtensor_adaptor<xt::xiterator_adaptor<xt::xiterator<xt::xstepper<xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major>, xt::xiterator<xt::xstepper<const xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major> >, 1, xt::layout_type::dynamic, xt::xtensor_expression_tag>' for 1st argument
        xtensor_adaptor& operator=(const xtensor_adaptor&);
                         ^
/Library/Frameworks/R.framework/Versions/3.6/Resources/library/xtensor/include/xtensor/xtensor.hpp:242:26: note: candidate function not viable: no known conversion from 'int' to 'xt::xtensor_adaptor<xt::xiterator_adaptor<xt::xiterator<xt::xstepper<xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major>, xt::xiterator<xt::xstepper<const xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major> >, 1, xt::layout_type::dynamic, xt::xtensor_expression_tag>' for 1st argument
        xtensor_adaptor& operator=(xtensor_adaptor&&);
                         ^
/Library/Frameworks/R.framework/Versions/3.6/Resources/library/xtensor/include/xtensor/xtensor.hpp:243:26: note: candidate function not viable: no known conversion from 'int' to 'xt::xtensor_adaptor<xt::xiterator_adaptor<xt::xiterator<xt::xstepper<xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major>, xt::xiterator<xt::xstepper<const xt::xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::layout_type::dynamic, xt::detail::inner_storage_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::column_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> > >, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true> *, xt::layout_type::column_major> >, 1, xt::layout_type::dynamic, xt::xtensor_expression_tag>::temporary_type' (aka 'xtensor_container<xt::uvector<int, std::__1::allocator<int> >, 1UL, (xt::layout_type)0, xt::xtensor_expression_tag>') for 1st argument
        xtensor_adaptor& operator=(temporary_type&&);
                         ^
/Library/Frameworks/R.framework/Versions/3.6/Resources/library/xtensor/include/xtensor/xtensor.hpp:246:26: note: candidate template ignored: could not match 'xexpression<type-parameter-0-0>' against 'int'
        xtensor_adaptor& operator=(const xexpression<E>& e);
                         ^
1 error generated.
make: *** [test.o] Error 1
```

</details>

--- a/include/xtensor/xmanipulation.hpp
+++ b/include/xtensor/xmanipulation.hpp
@@ -224,7 +224,7 @@ namespace xt
         using const_iterator = decltype(e.template cbegin<L>());
         using adaptor_type = xiterator_adaptor<iterator, const_iterator>;
         constexpr layout_type layout = std::is_pointer<iterator>::value ? L : layout_type::dynamic;
-        using type = xtensor_adaptor<adaptor_type, 1, layout, extension::get_expression_tag_t<E>>;
+        using type = xtensor_view<adaptor_type, 1, layout, extension::get_expression_tag_t<E>>;
         return type(adaptor_type(e.template begin<L>(), e.template cbegin<L>(), e.size()), { e.size() });
     }
 

--- a/include/xtensor/xoptional.hpp
+++ b/include/xtensor/xoptional.hpp
@@ -428,6 +428,26 @@ namespace xt
         };
     }
 
+    /***************************************
+     * xtensor_view extension for optional *
+     ***************************************/
+
+    namespace extension
+    {
+        template <class EC, std::size_t N, layout_type L>
+        struct xtensor_view_optional_traits : xtensor_optional_traits<EC, N, L>
+        {
+            using derived_type = xtensor_view<EC, N, L, xoptional_expression_tag>;
+        };
+
+        template <class EC, std::size_t N, layout_type L>
+        struct xtensor_view_base<EC, N, L, xoptional_expression_tag>
+        {
+            using traits = xtensor_view_optional_traits<EC, N, L>;
+            using type = xcontainer_optional_base<traits>;
+        };
+    }
+
     /************************************************
      * xfunction extension for optional expressions *
      ************************************************/

--- a/include/xtensor/xtensor.hpp
+++ b/include/xtensor/xtensor.hpp
@@ -206,7 +206,7 @@ namespace xt
     template <class EC, std::size_t N, layout_type L, class Tag>
     class xtensor_adaptor : public xstrided_container<xtensor_adaptor<EC, N, L, Tag>>,
                             public xcontainer_semantic<xtensor_adaptor<EC, N, L, Tag>>,
-                            public extension::xtensor_container_base_t<EC, N, L, Tag>
+                            public extension::xtensor_adaptor_base_t<EC, N, L, Tag>
     {
     public:
 
@@ -253,6 +253,124 @@ namespace xt
         const storage_type& storage_impl() const noexcept;
 
         friend class xcontainer<xtensor_adaptor<EC, N, L, Tag>>;
+    };
+
+    /****************************
+     * xtensor_view declaration *
+     ****************************/
+
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    class xtensor_view;
+
+    namespace extension
+    {
+        template <class EC, std::size_t N, layout_type L, class Tag>
+        struct xtensor_view_base;
+
+        template <class EC, std::size_t N, layout_type L>
+        struct xtensor_view_base<EC, N, L, xtensor_expression_tag>
+        {
+            using type = xtensor_empty_base;
+        };
+
+        template <class EC, std::size_t N, layout_type L, class Tag>
+        using xtensor_view_base_t = typename xtensor_view_base<EC, N, L, Tag>::type;
+    }
+
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    struct xcontainer_inner_types<xtensor_view<EC, N, L, Tag>>
+    {
+        using storage_type = std::remove_reference_t<EC>;
+        using reference = inner_reference_t<storage_type>;
+        using const_reference = typename storage_type::const_reference;
+        using size_type = typename storage_type::size_type;
+        using shape_type = std::array<typename storage_type::size_type, N>;
+        using strides_type = get_strides_t<shape_type>;
+        using backstrides_type = get_strides_t<shape_type>;
+        using inner_shape_type = shape_type;
+        using inner_strides_type = strides_type;
+        using inner_backstrides_type = backstrides_type;
+        using temporary_type = xtensor_container<temporary_container_t<storage_type>, N, L, Tag>;
+        static constexpr layout_type layout = L;
+    };
+
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    struct xiterable_inner_types<xtensor_view<EC, N, L, Tag>>
+        : xcontainer_iterable_types<xtensor_view<EC, N, L, Tag>>
+    {
+    };
+
+    /**
+     * @class xtensor_view
+     * @brief Dense multidimensional container adaptor with view
+     * semantics and fixed dimension.
+     *
+     * The xtensor_view class implements a dense multidimensional
+     * container adaptor with viewsemantics and fixed dimension. It
+     * is used to provide a multidimensional container semantic and a
+     * view semantic to stl-like containers.
+     *
+     * @tparam EC The closure for the container type to adapt.
+     * @tparam N The dimension of the view.
+     * @tparam L The layout_type of the view.
+     * @tparam Tag The expression tag.
+     * @sa xstrided_container, xcontainer
+     */
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    class xtensor_view : public xstrided_container<xtensor_view<EC, N, L, Tag>>,
+                         public xview_semantic<xtensor_view<EC, N, L, Tag>>,
+                         public extension::xtensor_view_base_t<EC, N, L, Tag>
+    {
+    public:
+
+        using container_closure_type = EC;
+
+        using self_type = xtensor_view<EC, N, L, Tag>;
+        using base_type = xstrided_container<self_type>;
+        using semantic_base = xview_semantic<self_type>;
+        using extension_base = extension::xtensor_adaptor_base_t<EC, N, L, Tag>;
+        using storage_type = typename base_type::storage_type;
+        using allocator_type = typename base_type::allocator_type;
+        using shape_type = typename base_type::shape_type;
+        using strides_type = typename base_type::strides_type;
+        using backstrides_type = typename base_type::backstrides_type;
+        using temporary_type = typename semantic_base::temporary_type;
+        using expression_tag = Tag;
+
+        xtensor_view(storage_type&& storage);
+        xtensor_view(const storage_type& storage);
+
+        template <class D>
+        xtensor_view(D&& storage, const shape_type& shape, layout_type l = L);
+
+        template <class D>
+        xtensor_view(D&& storage, const shape_type& shape, const strides_type& strides);
+
+        ~xtensor_view() = default;
+
+        xtensor_view(const xtensor_view&) = default;
+        xtensor_view& operator=(const xtensor_view&);
+
+        xtensor_view(xtensor_view&&) = default;
+        xtensor_view& operator=(xtensor_view&&);
+
+        template <class E>
+        self_type& operator=(const xexpression<E>& e);
+
+        template <class E>
+        disable_xexpression<E, self_type>& operator=(const E& e);
+
+    private:
+
+        container_closure_type m_storage;
+
+        storage_type& storage_impl() noexcept;
+        const storage_type& storage_impl() const noexcept;
+
+        void assign_temporary_impl(temporary_type&& tmp);
+
+        friend class xcontainer<xtensor_view<EC, N, L, Tag>>;
+        friend class xview_semantic<xtensor_view<EC, N, L, Tag>>;
     };
 
     /************************************
@@ -431,9 +549,9 @@ namespace xt
         return m_storage;
     }
 
-    /*******************
-     * xtensor_adaptor *
-     *******************/
+    /**********************************
+     * xtensor_adaptor implementation *
+     **********************************/
 
     /**
      * @name Constructors
@@ -541,6 +659,122 @@ namespace xt
     inline auto xtensor_adaptor<EC, N, L, Tag>::storage_impl() const noexcept -> const storage_type&
     {
         return m_storage;
+    }
+
+    /*******************************
+     * xtensor_view implementation *
+     *******************************/
+
+    /**
+     * @name Constructors
+     */
+    //@{
+    /**
+     * Constructs an xtensor_view of the given stl-like container.
+     * @param storage the container to adapt
+     */
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    inline xtensor_view<EC, N, L, Tag>::xtensor_view(storage_type&& storage)
+        : base_type(), m_storage(std::move(storage))
+    {
+    }
+
+    /**
+     * Constructs an xtensor_view of the given stl-like container.
+     * @param storage the container to adapt
+     */
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    inline xtensor_view<EC, N, L, Tag>::xtensor_view(const storage_type& storage)
+        : base_type(), m_storage(storage)
+    {
+    }
+
+    /**
+     * Constructs an xtensor_view of the given stl-like container,
+     * with the specified shape and layout_type.
+     * @param storage the container to adapt
+     * @param shape the shape of the xtensor_view
+     * @param l the layout_type of the xtensor_view
+     */
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    template <class D>
+    inline xtensor_view<EC, N, L, Tag>::xtensor_view(D&& storage, const shape_type& shape, layout_type l)
+        : base_type(), m_storage(std::forward<D>(storage))
+    {
+        base_type::resize(shape, l);
+    }
+
+    /**
+     * Constructs an xtensor_view of the given stl-like container,
+     * with the specified shape and strides.
+     * @param storage the container to adapt
+     * @param shape the shape of the xtensor_view
+     * @param strides the strides of the xtensor_view
+     */
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    template <class D>
+    inline xtensor_view<EC, N, L, Tag>::xtensor_view(D&& storage, const shape_type& shape, const strides_type& strides)
+        : base_type(), m_storage(std::forward<D>(storage))
+    {
+        base_type::resize(shape, strides);
+    }
+    //@}
+
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    inline auto xtensor_view<EC, N, L, Tag>::operator=(const xtensor_view& rhs) -> self_type&
+    {
+        base_type::operator=(rhs);
+        m_storage = rhs.m_storage;
+        return *this;
+    }
+
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    inline auto xtensor_view<EC, N, L, Tag>::operator=(xtensor_view&& rhs) -> self_type&
+    {
+        base_type::operator=(std::move(rhs));
+        m_storage = rhs.m_storage;
+        return *this;
+    }
+
+    /**
+     * @name Extended copy semantic
+     */
+    //@{
+    /**
+     * The extended assignment operator.
+     */
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    template <class E>
+    inline auto xtensor_view<EC, N, L, Tag>::operator=(const xexpression<E>& e) -> self_type&
+    {
+        return semantic_base::operator=(e);
+    }
+    //@}
+
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    template <class E>
+    inline auto xtensor_view<EC, N, L, Tag>::operator=(const E& e) -> disable_xexpression<E, self_type>&
+    {
+        std::fill(m_storage.begin(), m_storage.end(), e);
+        return *this;
+    }
+    
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    inline auto xtensor_view<EC, N, L, Tag>::storage_impl() noexcept -> storage_type&
+    {
+        return m_storage;
+    }
+
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    inline auto xtensor_view<EC, N, L, Tag>::storage_impl() const noexcept -> const storage_type&
+    {
+        return m_storage;
+    }
+
+    template <class EC, std::size_t N, layout_type L, class Tag>
+    inline void xtensor_view<EC, N, L, Tag>::assign_temporary_impl(temporary_type&& tmp)
+    {
+        std::copy(tmp.cbegin(), tmp.cend(), m_storage.begin());
     }
 
     /**

--- a/include/xtensor/xtensor_forward.hpp
+++ b/include/xtensor/xtensor_forward.hpp
@@ -137,6 +137,9 @@ namespace xt
     template <class EC, std::size_t N, layout_type L = XTENSOR_DEFAULT_LAYOUT, class Tag = xtensor_expression_tag>
     class xtensor_adaptor;
 
+    template <class EC, std::size_t N, layout_type L = XTENSOR_DEFAULT_LAYOUT, class Tag = xtensor_expression_tag>
+    class xtensor_view;
+
     template <std::size_t... N>
     class fixed_shape;
 

--- a/test/test_xstrided_view.cpp
+++ b/test/test_xstrided_view.cpp
@@ -908,7 +908,19 @@ namespace xt
         xt::xarray<int> x = {{1, 2, 3}, {4, 5, 6}};
         auto x_view = xt::strided_view(x, {xt::range(0, 1), xt::range(0, 1)});
         auto x_flat = xt::flatten<xt::layout_type::column_major>(x_view);
-        x_flat = {10};
-        EXPECT_EQ(x(0, 0), 10);
+        x_flat = 10;
+        xt::xarray<int> exp = {{10, 2, 3}, {4, 5, 6}};
+        EXPECT_EQ(x, exp);
+
+        auto x_view2 = xt::strided_view(x, {xt::range(0, 1), xt::range(0, 2)});
+        auto x_flat2 = xt::flatten<xt::layout_type::column_major>(x_view2);
+        x_flat2 = 15;
+        xt::xarray<int> exp2 = {{15, 15, 3}, {4, 5, 6}};
+        EXPECT_EQ(x, exp2);
+
+        xt::xarray<int> b = {20, 25};
+        x_flat2 = b;
+        xt::xarray<int> exp3 = {{20, 25, 3}, {4, 5, 6}};
+        EXPECT_EQ(x, exp3);
     }
 }

